tree/tree.go

```diff
@@ -0,0 +1,723 @@
+
+// DefaultKeyMap is the default set of key bindings for navigating and acting
+// upon the tree.
+var DefaultKeyMap = KeyMap{
+	Down: key.NewBinding(
+		key.WithKeys("down", "j", "ctrl+n"),
+		key.WithHelp("↓/j", "down"),
+	),
+	Up: key.NewBinding(
+		key.WithKeys("up", "k", "ctrl+p"),
+		key.WithHelp("↑/k", "up"),
+	),
+	Toggle: key.NewBinding(
+		key.WithKeys("enter"),
+		key.WithHelp("⏎", "toggle"),
+	),
+
+	// Toggle help.
+	ShowFullHelp: key.NewBinding(
+		key.WithKeys("?"),
+		key.WithHelp("?", "more"),
+	),
+	CloseFullHelp: key.NewBinding(
+		key.WithKeys("?"),
+		key.WithHelp("?", "close help"),
+	),
+
+	Quit: key.NewBinding(
+		key.WithKeys("q", "ctrl+c"),
+		key.WithHelp("q", "quit"),
+	),
+}
+
+// Model is the Bubble Tea model for this tree element.
+type Model struct {
+	showHelp bool
+	// OpenCharacter is the character used to represent an open node.
+	OpenCharacter string
+	// ClosedCharacter is the character used to represent a closed node.
+	ClosedCharacter string
+	// KeyMap encodes the keybindings recognized by the widget.
+	KeyMap KeyMap
+	// styles sets the styling for the tree
+	styles Styles
+	Help   help.Model
+
+	// Additional key mappings for the short and full help views. This allows
+	// you to add additional key mappings to the help menu without
+	// re-implementing the help component. Of course, you can also disable the
+	// list's help component and implement a new one if you need more
+	// flexibility.
+	AdditionalShortHelpKeys func() []key.Binding
+	AdditionalFullHelpKeys  func() []key.Binding
+
+	root   *Node
+	width  int
+	height int
+	// yOffset is the vertical offset of the selected node.
+	yOffset int
+}
+
+// Node is a a node in the tree
+// Node implements lipgloss's tree.Node
+type Node struct {
+	// tree is used as the renderer layer
+	tree *ltree.Tree
+
+	// yOffset is the vertical offset of the selected node.
+	yOffset int
+
+	// depth is the depth of the node in the tree
+	depth int
+
+	// isRoot is true for every Node which was added with tree.Root
+	isRoot bool
+	open   bool
+
+	// value is the root value of the node
+	value any
+
+	// TODO: expose a getter for this in lipgloss?
+	rootStyle lipgloss.Style
+
+	opts *itemOptions
+
+	// TODO: move to lipgloss.Tree?
+	size int
+}
+
+// IsSelected returns whether this item is selected.
+func (t *Node) IsSelected() bool {
+	return t.yOffset == t.opts.treeYOffset
+}
+
+// Depth returns the depth of the node in the tree.
+func (t *Node) Depth() int {
+	return t.depth
+}
+
+// Size returns the number of nodes in the tree.
+// Note that if a child isn't open, its size is 1
+func (t *Node) Size() int {
+	return t.size
+}
+
+// YOffset returns the vertical offset of the Node
+func (t *Node) YOffset() int {
+	return t.yOffset
+}
+
+type itemOptions struct {
+	openCharacter   string
+	closedCharacter string
+	treeYOffset     int
+}
+
+// Used to print the Node's tree
+// TODO: Value is not called on the root node, so we need to repeat the open/closed character
```

Comment:
Here I found I needed to implement the String() to repeat the open/close character because the `Value()` func isn't called on the root. Maybe this is an easy fix in lipgloss.